test_host_regexp_multiple_ptr_records causes race in OpenSSL `OPENSSL_cleanup`

Happened in [test_host_regexp_multiple_ptr_records](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80102&sha=latest&name_0=PR&name_1=Integration+tests+%28tsan%2C+1%2F6%29).

[Test case history](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIHJlcG9ydF91cmwtLSwgdGVzdF9jb250ZXh0X3JhdwpGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICAtLUFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzIElOICgnRkFJTCcsICdFUlJPUicsICdCUk9LRU4nKQogICAgQU5EIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDI1LTAxLTAxJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9ob3N0X3JlZ2V4cF9tdWx0aXBsZV9wdHJfcmVjb3JkcyUnCiAgICAtLUFORCB0ZXN0X2NvbnRleHRfcmF3IGxpa2UgJyVCbG9jayBzdHJ1Y3R1cmUgbWlzbWF0Y2glJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIERFU0MKCg==)

@thevar1able 

```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Write of size 1 at 0x721000000280 by main thread:
E               #0 pthread_rwlock_destroy <null> (clickhouse+0x889f42e) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #1 CRYPTO_THREAD_lock_free ci/tmp/build/./contrib/openssl/crypto/threads_pthread.c:141:5 (clickhouse+0x2514de1e) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #2 ossl_rand_ctx_free ci/tmp/build/./contrib/openssl/crypto/rand/rand_lib.c:502:5 (clickhouse+0x2511c36b) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #3 ossl_release_default_drbg_ctx ci/tmp/build/./contrib/openssl/crypto/context.c:511:9 (clickhouse+0x2501ebc9) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #4 ossl_rand_cleanup_int ci/tmp/build/./contrib/openssl/crypto/rand/rand_lib.c:108:5 (clickhouse+0x2511bc36) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #5 OPENSSL_cleanup ci/tmp/build/./contrib/openssl/crypto/init.c:424:5 (clickhouse+0x250cb57c) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #6 at_exit_callback_installed_at() crtstuff.c (clickhouse+0x88e4ce3) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #7 ossl_init_register_atexit ci/tmp/build/./contrib/openssl/crypto/init.c:110:9 (clickhouse+0x250cbcfa) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #8 ossl_init_register_atexit_ossl_ ci/tmp/build/./contrib/openssl/crypto/init.c:98:1 (clickhouse+0x250cbcfa)
E           
E             Previous atomic read of size 1 at 0x721000000280 by thread T1:
E               #0 pthread_rwlock_rdlock <null> (clickhouse+0x889f5ae) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #1 CRYPTO_THREAD_read_lock ci/tmp/build/./contrib/openssl/crypto/threads_pthread.c:93:9 (clickhouse+0x2514dd59) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #2 RAND_get0_primary ci/tmp/build/./contrib/openssl/crypto/rand/rand_lib.c:715:10 (clickhouse+0x2511bcc2) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #3 RAND_status ci/tmp/build/./contrib/openssl/crypto/rand/rand_lib.c:322:17 (clickhouse+0x2511be97) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #4 ossl_random openssl.c (clickhouse+0x22786fda) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #5 Curl_ssl_random <null> (clickhouse+0x2277c30f) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #6 Curl_rand_bytes <null> (clickhouse+0x2273c86c) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #7 Curl_socketpair <null> (clickhouse+0x2274abb7) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #8 Curl_resolver_getaddrinfo <null> (clickhouse+0x226ff75c) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #9 Curl_getaddrinfo <null> (clickhouse+0x22710390) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #10 Curl_resolv <null> (clickhouse+0x227114b9) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #11 Curl_resolv_timeout <null> (clickhouse+0x22711a0f) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #12 resolve_server url.c (clickhouse+0x22758087) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #13 Curl_connect <null> (clickhouse+0x22755cc7) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #14 multi_runsingle multi.c (clickhouse+0x22734890) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #15 curl_multi_perform <null> (clickhouse+0x227339df) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #16 curl_easy_perform <null> (clickhouse+0x226ef480) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #17 sentry__curl_send_task ci/tmp/build/./contrib/sentry-native/src/transports/sentry_transport_curl.c:225:19 (clickhouse+0x11dc2f02) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #18 worker_thread ci/tmp/build/./contrib/sentry-native/src/sentry_sync.c:262:9 (clickhouse+0x11db64f7) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E           
E             Location is heap block of size 56 at 0x721000000280 allocated by main thread:
E               #0 malloc <null> (clickhouse+0x889a9b0) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #1 CRYPTO_malloc ci/tmp/build/./contrib/openssl/crypto/mem.c:202:11 (clickhouse+0x250e4e54) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #2 CRYPTO_zalloc ci/tmp/build/./contrib/openssl/crypto/mem.c:222:11 (clickhouse+0x250e4e54)
E               #3 CRYPTO_THREAD_lock_new ci/tmp/build/./contrib/openssl/crypto/threads_pthread.c:50:17 (clickhouse+0x2514dcff) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #4 ossl_rand_ctx_new ci/tmp/build/./contrib/openssl/crypto/rand/rand_lib.c:475:18 (clickhouse+0x2511c2be) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #5 context_init ci/tmp/build/./contrib/openssl/crypto/context.c:110:17 (clickhouse+0x2501e4fd) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #6 default_context_do_init ci/tmp/build/./contrib/openssl/crypto/context.c:374:10 (clickhouse+0x2501eaaf) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #7 default_context_do_init_ossl_ ci/tmp/build/./contrib/openssl/crypto/context.c:369:1 (clickhouse+0x2501eaaf)
E               #8 pthread_once <null> (clickhouse+0x88a055d) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #9 CRYPTO_THREAD_run_once ci/tmp/build/./contrib/openssl/crypto/threads_pthread.c:152:9 (clickhouse+0x2514de60) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #10 get_thread_default_context ci/tmp/build/./contrib/openssl/crypto/context.c:397:10 (clickhouse+0x2501ec31) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #11 get_default_context ci/tmp/build/./contrib/openssl/crypto/context.c:405:36 (clickhouse+0x2501ec31)
E               #12 ossl_lib_ctx_get_concrete ci/tmp/build/./contrib/openssl/crypto/context.c:521:16 (clickhouse+0x2501ec31)
E               #13 ossl_lib_ctx_get_data ci/tmp/build/./contrib/openssl/crypto/context.c:548:11 (clickhouse+0x2501ec31)
E               #14 get_provider_store ci/tmp/build/./contrib/openssl/crypto/provider_core.c:331:13 (clickhouse+0x25117362) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #15 ossl_provider_disable_fallback_loading ci/tmp/build/./contrib/openssl/crypto/provider_core.c:341:18 (clickhouse+0x25117362)
E               #16 OSSL_PROVIDER_load_ex ci/tmp/build/./contrib/openssl/crypto/provider.c:61:9 (clickhouse+0x2511595d) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #17 OSSL_PROVIDER_load ci/tmp/build/./contrib/openssl/crypto/provider.c:68:12 (clickhouse+0x2511595d)
E               #18 DB::OpenSSLInitializer::initialize() ci/tmp/build/./src/Common/Crypto/OpenSSLInitializer.cpp:50:28 (clickhouse+0x1198b8fd) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #19 init_ssl() ci/tmp/build/./programs/main.cpp:238:5 (clickhouse+0x891fef3) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E           
E             Thread T1 (tid=8, running) created by main thread at:
E               #0 pthread_create <null> (clickhouse+0x889c4d1) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #1 sentry__bgworker_start ci/tmp/build/./contrib/sentry-native/src/sentry_sync.c:293:9 (clickhouse+0x11db6115) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #2 sentry__curl_transport_start ci/tmp/build/./contrib/sentry-native/src/transports/sentry_transport_curl.c:118:12 (clickhouse+0x11dc3535) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #3 sentry__transport_startup ci/tmp/build/./contrib/sentry-native/src/sentry_transport.c:91:18 (clickhouse+0x11db793d) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #4 sentry_init ci/tmp/build/./contrib/sentry-native/src/sentry_core.c:141:13 (clickhouse+0x11dac86f) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #5 SentryWriter::SentryWriter(Poco::Util::LayeredConfiguration&) ci/tmp/build/./src/Daemon/SentryWriter.cpp:113:27 (clickhouse+0x11d95cd5) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #6 SentryWriter::initializeInstance(Poco::Util::LayeredConfiguration&) ci/tmp/build/./src/Daemon/SentryWriter.cpp:69:38 (clickhouse+0x11d94dcd) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #7 BaseDaemon::initializeTerminationAndSignalProcessing() ci/tmp/build/./src/Daemon/BaseDaemon.cpp:416:5 (clickhouse+0x11d74eb4) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #8 BaseDaemon::initialize(Poco::Util::Application&) ci/tmp/build/./src/Daemon/BaseDaemon.cpp:402:5 (clickhouse+0x11d71c9f) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #9 DB::Server::initialize(Poco::Util::Application&) ci/tmp/build/./programs/server/Server.cpp:594:17 (clickhouse+0x11c88a43) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #10 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:310:2 (clickhouse+0x21d558b2) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #11 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:588:25 (clickhouse+0x11c888a5) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #12 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x21d732c0) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #13 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:385:20 (clickhouse+0x11c8550c) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
E               #14 main ci/tmp/build/./programs/main.cpp:304:21 (clickhouse+0x8920540) (BuildId: 230d5114e7047636618e9a07c8b58e0d6613a006)
```

### How to reproduce

_No response_

### Error message and/or stacktrace

_No response_

--- a/src/Common/Crypto/OpenSSLInitializer.cpp
+++ b/src/Common/Crypto/OpenSSLInitializer.cpp
@@ -2,12 +2,12 @@
 
 #include "config.h"
 
-#include <Common/OpenSSLHelpers.h>
 #include <Common/Exception.h>
+#include <Common/OpenSSLHelpers.h>
 
 #if USE_SSL
-#    include <openssl/provider.h>
 #    include <openssl/crypto.h>
+#    include <openssl/provider.h>
 #    include <openssl/ssl.h>
 #endif
 
@@ -31,6 +31,10 @@ void OpenSSLInitializer::initialize()
 #if USE_SSL
     if (ref_count++ == 0)
     {
+        // Disable OpenSSL atexit hook.
+        // It may cause issues on shutdown, when some OpenSSL objects are still in use.
+        auto openssl_flags = OPENSSL_INIT_NO_ATEXIT;
+
         /// Loading OpenSSL config only if it is set explicitly.
         ///
         /// 'legacy' provider may be disabled in system config,
@@ -38,14 +42,15 @@ void OpenSSLInitializer::initialize()
         const char * env_openssl_conf = getenv("OPENSSL_CONF"); // NOLINT(concurrency-mt-unsafe)
         if (env_openssl_conf)
         {
-            if (OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, nullptr) == 0)
-                throw std::runtime_error(fmt::format("Failed to load OpenSSL config. {}", getOpenSSLErrors()));
+            openssl_flags |= OPENSSL_INIT_LOAD_CONFIG;
         }
         else
         {
-            if (OPENSSL_init_ssl(OPENSSL_INIT_NO_LOAD_CONFIG, nullptr) == 0)
-                throw std::runtime_error(fmt::format("Failed to initialize OpenSSL. {}", getOpenSSLErrors()));
+            openssl_flags |= OPENSSL_INIT_NO_LOAD_CONFIG;
         }
+
+        if (OPENSSL_init_ssl(openssl_flags, nullptr) == 0)
+            throw std::runtime_error(fmt::format("Failed to load OpenSSL config. {}", getOpenSSLErrors()));
 
         default_provider = OSSL_PROVIDER_load(nullptr, "default");
         if (!default_provider)

